tensorboard/java/org/tensorflow/tensorboard/vulcanize/Vulcanize.java

```diff
@@ -778,6 +778,79 @@ private static String extractAndTransformJavaScript(Document document, Webpath j
     return scriptContent;
   }
 
+  private static void cloneChildrenWithoutWhitespace(Element src, Element dest) {
+    List<Node> toMove = new ArrayList<Node>();
+    for (Node node : src.childNodes()) {
+      if (node instanceof TextNode && ((TextNode) node).isBlank()) {
+        continue;
+      }
+      toMove.add(node);
+    }
+    for (int i = 0; i < toMove.size(); i++) {
+      Node node = toMove.get(i);
+      dest.appendChild(node.clone());
+    }
+  }
+
+  // When we inline the HTML based on `<link rel="import">` in `transform`, we
+  // replace the link element with parsed document. This makes us have nested
+  // documents and jsoup's Node.outerHtml (or Node.toString) are incapable of
+  // properly outputting that. Here, we flatten the document by combining all
+  // elements in `<head>` and `<body>` of nested document in one `<head>` and
+  // `<body>`.
+  //
+  // Examples:
+  // // Input
+  // <#root> <!-- document -->
+  //   <html>
+  //     <head><#root><html><body>foo</body></html></#root></head>
+  //     <body><span>bar</span></body>
+  //   </html>
+  // </html>
+  // // Output
+  // <#root> <!-- document -->
+  //   <html>
+  //     <head></head>
+  //     <body>foo<span>bar</span></body>
+  //   </html>
+  // </html>
+  private static Document getFlattenedDocument(Document document) {
+    Document flatDoc = new Document("/");
+    flatDoc.normalise();
+    Element rootDocumentHead = flatDoc.head();
+    Element rootDocumentBody = flatDoc.body();
+
+    Node currentNode = document;
+    while (currentNode != null) {
+      // Do not clone the element if it is a `document` inside `<head>`.
```

Comment:
Sorry, I feel like all my comments are just about documentation. (Which I guess is good - the code itself seems fine once you understand what it's trying to do).

This comment is mostly about what the next bit of code *does not do* but it probably would be clearer if rewritten so that it first explains what it *does do*.  It took me a while to figure it out.

Namely:

Copy most children of \<head\> elements into the flat doc without modification. The exception is if the node is of type `document` (or a <#root> element) in which case we want to traverse further and only copy the elements in its `body` and `head` elements.